### ISSUE TITLE
if you've got time: "XMLException"

### ISSUE THREAD

**Issue by NickSzen** [Designer225/CustomTroopUpgrades#8]
Hi, I'm very new to this, and I don't quite see what I've screwed up. If you've got the time and patience, I'd appreciate the help.  Thanks for the mod and your time.
![MountAndBladeBannerlord crash capture](https://user-images.githubusercontent.com/64409042/80362497-bdb8fd80-8850-11ea-9754-78190f9fa475.PNG)

[SubModule - Copy.txt](https://github.com/Designer225/CustomTroopUpgrades/files/4539068/SubModule.-.Copy.txt)
[spnpccharacters_BattanianSpecialists.txt](https://github.com/Designer225/CustomTroopUpgrades/files/4539072/spnpccharacters_BattanianSpecialists.txt)
[BattanianSpecialistsReplacement.txt](https://github.com/Designer225/CustomTroopUpgrades/files/4539073/BattanianSpecialistsReplacement.txt)

I'm trying to modify a few Battanian units and the Freebooter Forest Bandit.  
My aim is to turn the Skene, Kern, Redshank tree into a heavy shield infantry and to make the tree more accessible without cheating/mods by attaching the Skene to the Freebooter; replacing the Falxman upgrade path.
The last thing I included was a few tweaks to the Fian Champion: slightly heavier armor, the two-handed falx instead of the one-handed sword they currently have, and a few very minor stat changes.
Thanks again.

**Comment by Designer225** [comment]
Try removing the Fian Champions to see if the issue persists. I need to see what did the game try to load that caused the crash.

**Comment by NickSzen** [comment]
Okay. The issue persists:
(The number on the file is just for organization of documentation. I haven't changed the XML file names)
[SubModule(2).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4543467/SubModule.2.txt)
[spnpccharacters_BattanianSpecialists(2).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4543480/spnpccharacters_BattanianSpecialists.2.txt)
[bandits_BattanianSpecialists(2).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4543483/bandits_BattanianSpecialists.2.txt)
[BattanianSpecialistsReplacement(2).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4543484/BattanianSpecialistsReplacement.2.txt)
and the results:
![MountAndBladeBannerlord crash capture 2a- 8388575 tag](https://user-images.githubusercontent.com/64409042/80450148-e1328580-88ee-11ea-8351-324fb0d87923.PNG)
I changed the replace tag to see if that did anything, but no:
![MountAndBladeBannerlord crash capture 2b-  8388607 tag](https://user-images.githubusercontent.com/64409042/80450201-01fadb00-88ef-11ea-981e-6d46a95a4d08.PNG)

I fiddled with the load order a bit too, moving things, turning things off, removing them; but got no change in results:
![MountAndBladeBannerlord load order capture](https://user-images.githubusercontent.com/64409042/80450249-1d65e600-88ef-11ea-884e-69a484c7f44f.PNG)

**Comment by NickSzen** [comment]
I then isolated the Gallowglass tree (Skene, Kern, Redshank) from the Freebooter upgrade and the bandit NPCCharacters XML:
[SubModule(3).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4543509/SubModule.3.txt)
[spnpccharacters_BattanianSpecialists(3).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4543513/spnpccharacters_BattanianSpecialists.3.txt)
[BattanianSpecialistsReplacement(3).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4543514/BattanianSpecialistsReplacement.3.txt)
with no change using either "everything" tag:
![MountAndBladeBannerlord crash capture 3a- 8388575 tag](https://user-images.githubusercontent.com/64409042/80450659-0378d300-88f0-11ea-9013-1723e3ecbd89.PNG)
![MountAndBladeBannerlord crash capture 3b- 8388607 tag](https://user-images.githubusercontent.com/64409042/80450669-083d8700-88f0-11ea-82d8-a185bad91ee5.PNG)

**Comment by NickSzen** [comment]
Finally, I reversed it, and isolated the Fian Champions:
*used same SubModule XML as (3)
[spnpccharacters_BattanianSpecialists(4).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4543524/spnpccharacters_BattanianSpecialists.4.txt)
[BattanianSpecialistsReplacement(4).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4543525/BattanianSpecialistsReplacement.4.txt)
same error again, both tags:
![MountAndBladeBannerlord crash capture 4- 8388607 tag](https://user-images.githubusercontent.com/64409042/80450896-8863ec80-88f0-11ea-936c-7eb4e1821c94.PNG)

So far as my limited understanding of the error message carries me, it's a merge issue. It mentions "GameMenusCallbacks". Could this be a mod conflict and not my bad writing? Maybe MBOptionScreen? I already checked the Developer Console and CITEditor. Thought about this as I was writing. I'll check this now.

**Comment by NickSzen** [comment]
Nope. Or not exactly.
Removing my modules one at a time, MBOptionScreen and its update, Mod Configuration Menu, come and go without change.
ModLib represents an issue, though. I don't know what exactly that issue is, but ModLib seems to be a factor.
With ModLib active, I get the above "InvalidOperationException".  With ModLib out of my module folder, I get a full crash.
As I'm typing this, I realize I should have tested if the game will load after removing ModLib and my module. But I'm tired now, so I'm going to bed and I'll probably try that later today.
edit: including crash logs
[rgl_log_28312.txt](https://github.com/Designer225/CustomTroopUpgrades/files/4544076/rgl_log_28312.txt)
[rgl_log_errors_28312.txt](https://github.com/Designer225/CustomTroopUpgrades/files/4544079/rgl_log_errors_28312.txt)

**Comment by Designer225** [comment]
Maybe these unintentional comments?

```xml
  <!--NPCCharacter id="galloglass_tier_1_new"
```
and
```xml
  <!--NPCCharacter id="battanian_fian"
```
There also appears to be a comment that isn't properly closed (in the original):
```xml
<!-- Adjusted Battanian Fian Champion stats and equipment >
<stuff>
```

**Comment by NickSzen** [comment]
We're now working with 2 minor issues:
[SubModule(working01).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4548517/SubModule.working01.txt)
[spnpccharacters_BattanianSpecialists(working 01).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4548524/spnpccharacters_BattanianSpecialists.working.01.txt)
[BattanianSpecialistsReplacement(working 01).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4548526/BattanianSpecialistsReplacement.working.01.txt)

The stats and equipment changed correctly, the upgrade functions correctly, everything operates in game and in battle normally with desired changes, but there are two Battanian Fian Champions:
![MountAndBladeBannerlord working 01](https://user-images.githubusercontent.com/64409042/80539128-f81bbb00-8974-11ea-8ef3-74f5f8f15b66.PNG)
And two encyclopedia entries:
![MountAndBladeBannerlord working 02 double encyclopedia entry](https://user-images.githubusercontent.com/64409042/80539167-0c5fb800-8975-11ea-8f86-ae48a59916e1.PNG)
Both units have the updated stats and equipment, and both are viable in-game and cause no operational issue.  The only perceivable difference between the units is one is slightly smaller:
![MountAndBladeBannerlord working 03 slightly smaller character dimensions](https://user-images.githubusercontent.com/64409042/80539377-65c7e700-8975-11ea-8357-fa6c9a253c2b.PNG)
and is properly included in the troop tree:
![MountAndBladeBannerlord working 03 in upgrade tree](https://user-images.githubusercontent.com/64409042/80539430-7c6e3e00-8975-11ea-857a-14ea958a5a1b.PNG)

The second is the original size of the unmodified unit:
![MountAndBladeBannerlord working 02 original character dimensions](https://user-images.githubusercontent.com/64409042/80539592-c35c3380-8975-11ea-9463-591f6eedaf33.PNG)
and is outside of the troop trees and has a singular encyclopedia entry:
![MountAndBladeBannerlord working 02 outside of upgrade trees](https://user-images.githubusercontent.com/64409042/80539663-e7b81000-8975-11ea-8268-6f1f5b39841a.PNG)

Without CITEditor, the stand alone unit is inaccessible, so it just represents a little bit of messiness. 
Any idea as to the cause?
The size issue is a result of the Detailed Character Creator mod and is just another bit of messiness.

I'm going to check viability for other units in a few hours.
Thanks for your help.

**Comment by Designer225** [comment]
The duplicate troop is the "side effect" of working around the bane of every Bannerlord troop mod's existence. Basically, if you only overwrite a troop's definition, the game insists on merging the original definition with the modded one, and this can severely break immersion (e.g. vanilla equipment). Modders work around this by defining modified copies of the same troops and reconfigure the upgrade paths with the copies. That itself runs into issue of other troop trees being left behind, which is one of the things my mod attempts to address. Anyway, the workaround is the source of the duplicate troop issue, above. It's either that or merged definition, and most modders would take the former over the latter.

In other words, the above is "working as intended." And it's not the only game issue that is a "feature."

**Comment by NickSzen** [comment]
Thanks for clearing that up.
I've been messing with and testing the code for a few hours now, and I've got my modified heavy infantry in the game and working, but I can't use the Gallowglass tree to it.  I ended up using the Wolfskin tree instead.  The issue, as near as I can determine from my testing, is that the game doesn't like you changing the skill template between classes.  The Gallowglass are skirmishers originally, and I wasn't able to find any other inconsistency or error based on my tests with the Wolfskins whose Xml info features the same data points and architecture but start as heavy infantry for their skill template.
I could be wrong.
I'll try to add the Freebooter tie-in later today.
Another quick question: how does the name id work?
name="{=1zObCG2o}Battanian Fian Champion"
How do i translate a name ("Battanian Fian Champion") to the string ("=1zObCG2o")
Thanks again for your help. This has been fun.

**Comment by Designer225** [comment]
The "string" is the string id of the name, and is used for localization (i.e. other languages have different translations). I think the "name" is the default if there is no matching string. Since you are editing vanilla units, you don't need to worry about it.

**Comment by NickSzen** [comment]
I had done a quick search about character ids/names and I saw that the {=*} referenced a strings.txt, but I had no idea what exactly was going on there, so thanks for the explanation.  I was asking because I felt like I was getting close to getting my module fully operational and I was getting ambitious. I started planning out something a little more expansive, and my plan to mess with the Imperial troop tree involves referencing some different historical units than the ones used in game.

And  now I do have a fully operational module with all my initial plans realized, if with a slight twist.  The game really didn't like me editing the type of skill template, so I now have incorporated higher level Gallowglass skirmishers as a secondary upgrade branch off the Freebooter with the Fian and Fian Champion.  The higher level skirmisher works better for game balance and interesting army construction as well as better reflecting actual historical Celtic armies and tactics.  However this is getting a little pedantic, so I'll stop boring you now.

Here are the files if you want to look it over:
[SubModule(working base).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4555749/SubModule.working.base.txt)
[spnpccharacters_BattanianSpecialists(working base).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4555752/spnpccharacters_BattanianSpecialists.working.base.txt)
[BattanianSpecialistsReplacement(working base).txt](https://github.com/Designer225/CustomTroopUpgrades/files/4555753/BattanianSpecialistsReplacement.working.base.txt)

I'm going to play around with this for a bit, but I've got an idea for tweaks to the Imperials, Sturgians, and Vlandians to balance the buffed higher end Battanians.  I'm still working on how to incorporate the Khuzaits and Asserai, but I'll get there.

Thanks again for your help, your mod, and the seed for this idea.
